### DISCUSSION TITLE
Single table inheritance typing overrides

### DISCUSSION THREAD

**Discussion by sjaq** [sqlalchemy/sqlalchemy#9978] [Usage Questions] [answered]
I want to create a model that depending on it's status returns a different class with different typing information.

The setup I'm aiming for is simply that there is a "DRAFT" version of the model in which a lot of the columns are optional, when these fields are populated the model transitions to the "COMPLETE" status for which all the typing information is updated to exclude the `| None` type annotation. I'm trying to bind this directly into SQLAlchemy as it would lower the complexity of the code considerably and would be really nice to avoid checking if all the optional types are populated all over the codebase.

A setup like this seems to work:

```py
from enum import Enum

from sqlalchemy import create_engine
import sqlalchemy as sq
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, Session

engine = create_engine("postgresql+psycopg://sq:sq@localhost/sq_development", echo=True)


class Base(DeclarativeBase):
    pass


class ProjectStatus(str, Enum):
    DRAFT = "DRAFT"
    COMPLETE = "COMPLETE"


class Project(Base):
    __tablename__ = "project"

    id: Mapped[int] = mapped_column(primary_key=True)
    status: Mapped[ProjectStatus] = mapped_column(
        sq.Enum(ProjectStatus, name="project_status"),
        nullable=False,
        default=ProjectStatus.DRAFT,
    )

    title: Mapped[str | None] = mapped_column(nullable=True)

    __mapper_args__ = {"polymorphic_on": "status"}


class ProjectDraft(Project):
    __mapper_args__ = {"polymorphic_identity": ProjectStatus.DRAFT}


class ProjectComplete(Project):
    title: Mapped[str] = mapped_column(use_existing_column=True, nullable=True)
    # OR:
    title: Mapped[str] = Project.title

    __mapper_args__ = {"polymorphic_identity": ProjectStatus.COMPLETE}


# Create tables
Base.metadata.drop_all(engine)
Base.metadata.create_all(engine)

with Session(bind=engine) as s:
    s.add(Project(id=1, title="Something", status=ProjectStatus.COMPLETE))

    assert s.get(Project, 1).__class__ == ProjectComplete
```

I'm just wondering about the following:
- Is this use-case something that is supported?
- Maybe there is an even nicer way of doing this that I'm overlooking?
- For the alternative (more restrictive) version of the model I'm noticing there's two ways to reference the original column, by directly passing `ParentModel.column_name` or by passing `use_existing_column`, the only benefit of using the former is that I don't have to specify all the column options again, but I'm not sure if this API is really supported?

**Comment by zzzeek** [comment, marked as answer]
sure, what you have above there is fine, the typing hints you have like for "title" are only read by SQLAlchemy when the mappings are first set up, so if you are using standard polymorphic schemes to get back ProjectComplete objects, that's your class and that's the typing information you'll get.

an alternate way to set up that column would be to use TYPE_CHECKING

```py
from typing import TYPE_CHECKING

class ProjectComplete(Project):
    if TYPE_CHECKING:
        title: Mapped[str]

    __mapper_args__ = {"polymorphic_identity": ProjectStatus.COMPLETE}


```

that's likely a little safer than the `title: Mapped[str] = Project.title` approach, which I'm not as sure about.   use_existing_column=True is also safe.

**Reply by sjaq** [reply]
Thanks for the super fast reply! The `TYPE_CHECKING` approach you're mentioning here indeed seems like the safest setup, as the only thing I'm basically overriding is the type annotations it makes the most sense to me. Thanks!

**Comment by sjaq** [comment]
@zzzeek Another issue I'm running into with this approach is how to deal with the status changing, I see this error in the logs when I change the status:

```
SAWarning: Flushing object <ProjectDraft at 0x10705de40> with incompatible polymorphic identity <ProjectStatus.COMPLETE: 'COMPLETE'>; the object may not refresh and/or load correctly (this warning may be suppressed after 10 occurrences)
```

Is this something I can safely ignore if I consistently use the base class `Project` to retrieve the instances from the session?

**Reply by zzzeek** [reply]
ideally you wouldn't change the value of that column in-place, because if you have `ProjectDraft(status=COMPLETE)`, that's technically an invalid state

I guess from that we can derive a more complete typing model as far as what the ORM assumes

```py
class ProjectDraft(Project):
    __mapper_args__ = {"polymorphic_identity": ProjectStatus.DRAFT}

    if TYPE_CHECKING:
        status: Mapped[Literal[ProjectStatus.DRAFT]]


class ProjectComplete(Project):
    if TYPE_CHECKING:
        status: Mapped[Literal[ProjectStatus.COMPLETE]]
        title: Mapped[str]

    __mapper_args__ = {"polymorphic_identity": ProjectStatus.COMPLETE}

```

**Reply by zzzeek** [reply]
that is, you would replace the object in an ideal world 

```py
proj_complete = ProjectComplete(... fields from project_draft ...)
session.delete(proj_draft)
session.add(proj_complete)
```
you would see the ORM do an UPDATE for that row

**Reply by sjaq** [reply]
Ah I see, but then it would be impossible to transition from one state to another given an existing model in the session?

Or is there still a way to do this "the right way"?

**Reply by sjaq** [reply]
Was too quick, thanks, I'll look into this

**Reply by sjaq** [reply]
I can see that the `delete` then `add` approach works, but triggers an update to all columns in the model, which is a little less then ideal. But I guess I can find a way to make this work.

Would be really great if there was an API to transition a STI model from one base class to another, from my old rails days I can remember using something like this: https://apidock.com/rails/ActiveRecord/Persistence/becomes

**Reply by zzzeek** [reply]
you can also emit an UPDATE directly

```py
session.expunge(project_draft)
project_complete = session.scalars(update(Project).values(status=ProjectStatus.COMPLETE).where(Project.id == id).returning(ProjectComplete)).one()
```

**Reply by sjaq** [reply]
Thanks, I'll probably use an approach like that as it feels the least "hacky"
